plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/loops/SelfHyperLoopLabels.java

```diff
@@ -0,0 +1,257 @@
+import org.eclipse.elk.alg.layered.intermediate.LabelManagementProcessor;
+import org.eclipse.elk.alg.layered.options.LayeredOptions;
+import org.eclipse.elk.core.labels.ILabelManager;
+import org.eclipse.elk.core.math.KVector;
+import org.eclipse.elk.core.options.Direction;
+import org.eclipse.elk.core.options.PortSide;
+
+/**
+ * Represents the labels associated with a {@link SelfHyperLoop}. The labels are gathered as edges are added to the
+ * hyper loop. Their actual position is determined by a port side and an alignment on that side.
+ */
+public class SelfHyperLoopLabels {
+    
+    /**
+     * Describes all the different alignments a label might have. Most are alignments for labels on the northern or
+     * southern side. Labels on the eastern or western side are restricted to {@link Alignment#TOP}. 
+     */
+    public static enum Alignment {
+        /** A northern or southern centered label. */
+        CENTER,
+        /** A northern or southern left-aligned label. */
+        LEFT,
+        /** A northern or southern right-aligned label. */
+        RIGHT,
+        /** An eastern or western top-aligned label. */
+        TOP
+    }
+
+    // CHECKSTYLEOFF VisibilityModifier
+    /** An ID that can be used arbitrarily. */
+    public int id = 0;
+    // CHECKSTYLEON VisibilityModifier
+    
+    /** The labels represented by this instance. */
+    private final List<LLabel> lLabels = new ArrayList<>();
+    /** The size required to place all labels. */
+    private final KVector size = new KVector();
+    /** The position of our bounding box's top left corner. */
+    private final KVector position = new KVector();
+    /** The graph's layout direction. Influences the way the labels will be placed. */
+    private final Direction layoutDirection;
+    /** Space to leave between adjacent labels. */
+    private final double labelLabelSpacing;
+    
+    /** The side the label is placed on. */
+    private PortSide side;
+    /** The label's alignment. */
+    private Alignment alignment;
+    /** The port any non-center alignment is relative to. */
+    private SelfLoopPort alignmentReferenceSLPort;
+    
+    
+    /**
+     * Creates a new instance for the given {@link SelfHyperLoop}.
+     */
+    public SelfHyperLoopLabels(final SelfHyperLoop slLoop) {
+        // Initialize our properties from the graph
+        LNode lNode = slLoop.getSLHolder().getLNode();
+        
+        layoutDirection = lNode.getGraph().getProperty(LayeredOptions.DIRECTION);
+        labelLabelSpacing = LGraphUtil.getIndividualOrInherited(lNode, LayeredOptions.SPACING_LABEL_LABEL);
+    }
+
+    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // LLabel Access
+    
+    /**
+     * Adds the given collection of labels to this instance.
+     */
+    public void addLLabels(final Collection<LLabel> newLLabels) {
+        for (LLabel newLLabel : newLLabels) {
```

Comment:
Is there a way to influence the order of the labels being added? I.e. can the user somehow decide which label shall be, e.g., the top-most?

Reply:
Depends on what you mean by "user". The order is simply the same order the original ElkLabels are in, so the application developer can influence that order. As a consequence they can decide to give their users control over the order as well.